### DISCUSSION TITLE
Instructions to use MicroPython and font-to-py on a Pico powered Waveshare e-paper

### DISCUSSION THREAD

**Discussion by CoenTempelaars** [micropython/micropython#9157] [Show and tell]
I have written a small instructable how to use the [Writer class from micropython-font-to-py](https://github.com/peterhinch/micropython-font-to-py/tree/master/writer) on a Pico powered 3.7 inch Waveshare e-paper, rotate the display 90 degrees and center the text horizontally and vertically.

https://www.instructables.com/Waveshare-E-paper-Display-With-Raspberry-Pi-Pico-M

**Comment by jimmo** [comment]
That's really cool! A 2.7" one of these literally arrived in the mail today, so keen to give this a go.

**Comment by peterhinch** [comment]
That is great!

You've actually done almost all of the work necessary to run [nano-gui]() if you want to do more than display text. This will let you draw various controls, plot graphs and suchlike as per this demo:

![Image](https://github.com/peterhinch/micropython-nano-gui/raw/master/images/seismo.JPG)

**Comment by peterhinch** [comment]
Looking at your display driver I'm puzzled at the way you do rotation. Your approach allocates a 16,800 byte [buffer](https://github.com/CoenTempelaars/raspi-pico-epaper/blob/4708537cc699b162056fbc58ddc3c00af674bf78/python/epd_3in7.py#L222) every time you execute `show`. This runs the risk of an allocation failure after a period of running. One option is to allocate the buffer in `__init__`. But it seems to me that the buffer could be eliminated, with the rotation being done [here](https://github.com/CoenTempelaars/raspi-pico-epaper/blob/4708537cc699b162056fbc58ddc3c00af674bf78/python/epd_3in7.py#L246) as the data is output.

You've done a great job trimming Waveshare's wordy code, but there is scope for a further haircut. To give one example the `.digital_write` method doesn't do much and its calls could be treated thus:
```python
self.digital_write(self.reset_pin, 1)
```
could be reduced to
```python
self.reset_pin(1)
```

**Comment by CoenTempelaars** [comment]
Thanks for the feedback!

I'm going to start tinkering with nano-gui once I'm done with printing text in pretty fonts.

I like the idea of applying the rotation "on the fly" without the need for a separate buffer. I'm going to look into that, but that will definitely require some puzzling. If I fail to rotate the buffer "on the fly" then I'll change the code to allocate the extra buffer in `__init__`

I was not aware that Pins are callable in the way you describe. I'm going to remove the `.digital_read` and `.digital_write` methods once I figure out how to do that properly.

**Reply by peterhinch** [reply]
As a general comment I have just started porting the driver for the [Waveshare 4.2 inch ePaper](https://thepihut.com/collections/epaper-displays-for-raspberry-pi/products/4-2-e-paper-display-module-for-raspberry-pi-pico-black-white-400x300) display. I spotted bugs in the first few minutes of looking at the Waveshare driver. Further, they evidently have no grasp of optimising MicroPython code. The more I look at their code, the less I like it. Perhaps someone can explain what can possibly be the point of
```python
temp3 |= 0x00
```
because their driver is littered with similar...
